Add benchmark for performance comparison

taivu1998 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#43221 Add benchmark for performance comparison**
* #42867 Replace Conv1d with Conv2d

Differential Revision: [D23197567](https://our.internmc.facebook.com/intern/diff/D23197567)

Commit:
Add benchmark for performance comparison

[ghstack-poisoned]

taivu1998 commented:
Example: https://www.internalfb.com/intern/paste/P139226521/

kimishpatel commented on aten/src/ATen/benchmarks/stateful_conv1d.cpp:
```diff
@@ -0,0 +1,85 @@
+        for (size_t batch_size = 1; batch_size < 5; ++batch_size) {
+          for (size_t width = 32; width < 256; width *= 2) {
+            b->Args({input_channels, output_channels, kernel, batch_size, width, true});
+            b->Args({input_channels, output_channels, kernel, batch_size, width, false});
```
I think you want two benchmark, no? Otherwise how are you distinguish the reported results between optimized vs unoptimized? Looking at the <state> values reported by benchmark? I think that is fine as well.

taivu1998 replied:
So currently if we look at two consecutive lines, we can see the results for the two cases: optimized (Optimzed:1) and unoptimized(Optimzed:0). I think that way we can easily compare the performance for a given set of hyperparameter. Let me know if that's ok.

kimishpatel replied:
Sure, thats ok.

kimishpatel reviewed (changes requested):
Looks good. Can you add the benchmark results here?

kimishpatel commented:
> https://www.internalfb.com/intern/paste/P139226521/

This is internal fb link. Can you please remove this and instead paste the results directly?

kimishpatel commented:
Results are looking good. I am wondering if you can summarize the average improvement.

taivu1998 commented:
Example Run:

`------------------------------------------------------------------------------------------------------------------------------------------
Benchmark                                                                                                   Time           CPU Iterations
------------------------------------------------------------------------------------------------------------------------------------------
stateful_conv1d/Input Channels:32/Output Channels:32/Kernel:3/Batch Size:1/Width:32/Optimized:1       1075866 ns    1073923 ns        583
stateful_conv1d/Input Channels:32/Output Channels:32/Kernel:3/Batch Size:1/Width:32/Optimized:0       1465880 ns    1461682 ns        412
stateful_conv1d/Input Channels:32/Output Channels:32/Kernel:3/Batch Size:1/Width:64/Optimized:1       1326483 ns    1315410 ns        526
stateful_conv1d/Input Channels:32/Output Channels:32/Kernel:3/Batch Size:1/Width:64/Optimized:0       1873766 ns    1867842 ns        374
stateful_conv1d/Input Channels:32/Output Channels:32/Kernel:3/Batch Size:1/Width:128/Optimized:1      1904455 ns    1900167 ns        371
stateful_conv1d/Input Channels:32/Output Channels:32/Kernel:3/Batch Size:1/Width:128/Optimized:0      2969526 ns    2964561 ns        228
stateful_conv1d/Input Channels:32/Output Channels:32/Kernel:3/Batch Size:2/Width:32/Optimized:1       1554585 ns    1443135 ns        492
stateful_conv1d/Input Channels:32/Output Channels:32/Kernel:3/Batch Size:2/Width:32/Optimized:0      12660235 ns   12639436 ns         55
stateful_conv1d/Input Channels:32/Output Channels:32/Kernel:3/Batch Size:2/Width:64/Optimized:1       1929899 ns    1925342 ns        364
stateful_conv1d/Input Channels:32/Output Channels:32/Kernel:3/Batch Size:2/Width:64/Optimized:0      24904124 ns   24837338 ns         28
stateful_conv1d/Input Channels:32/Output Channels:32/Kernel:3/Batch Size:2/Width:128/Optimized:1      3173698 ns    3163716 ns        221
stateful_conv1d/Input Channels:32/Output Channels:32/Kernel:3/Batch Size:2/Width:128/Optimized:0     49094797 ns   48977970 ns         13
stateful_conv1d/Input Channels:32/Output Channels:32/Kernel:3/Batch Size:3/Width:32/Optimized:1       1691385 ns    1687278 ns        408
stateful_conv1d/Input Channels:32/Output Channels:32/Kernel:3/Batch Size:3/Width:32/Optimized:0      15167404 ns   15124956 ns         46
stateful_conv1d/Input Channels:32/Output Channels:32/Kernel:3/Batch Size:3/Width:64/Optimized:1       2573011 ns    2566797 ns        272
stateful_conv1d/Input Channels:32/Output Channels:32/Kernel:3/Batch Size:3/Width:64/Optimized:0      30213907 ns   30136943 ns         24
stateful_conv1d/Input Channels:32/Output Channels:32/Kernel:3/Batch Size:3/Width:128/Optimized:1      4408674 ns    4399762 ns        159
stateful_conv1d/Input Channels:32/Output Channels:32/Kernel:3/Batch Size:3/Width:128/Optimized:0     55516512 ns   55424121 ns         13
stateful_conv1d/Input Channels:32/Output Channels:32/Kernel:3/Batch Size:4/Width:32/Optimized:1       1980652 ns    1973277 ns        323
stateful_conv1d/Input Channels:32/Output Channels:32/Kernel:3/Batch Size:4/Width:32/Optimized:0      15965167 ns   15921340 ns         44
stateful_conv1d/Input Channels:32/Output Channels:32/Kernel:3/Batch Size:4/Width:64/Optimized:1       3265120 ns    3256765 ns        214
stateful_conv1d/Input Channels:32/Output Channels:32/Kernel:3/Batch Size:4/Width:64/Optimized:0      33907082 ns   33729891 ns         21
stateful_conv1d/Input Channels:32/Output Channels:32/Kernel:3/Batch Size:4/Width:128/Optimized:1      6444808 ns    6428044 ns        106
stateful_conv1d/Input Channels:32/Output Channels:32/Kernel:3/Batch Size:4/Width:128/Optimized:0     64191689 ns   64027354 ns         10`

taivu1998 commented:
| Benchmark | Time | CPU | Iterations |
| ----------- | ----- | ----- | ----------|
| stateful_conv1d/Input Channels:32/Output Channels:32/Kernel:3/Batch Size:1/Width:32/Optimized:1 | 1075866 ns | 1073923 ns | 583 |
| stateful_conv1d/Input Channels:32/Output Channels:32/Kernel:3/Batch Size:1/Width:32/Optimized:0 | 1465880 ns  |  1461682 ns    |    412 |
| stateful_conv1d/Input Channels:32/Output Channels:32/Kernel:3/Batch Size:2/Width:32/Optimized:1   |    1554585 ns    1443135 ns    |    492|
| stateful_conv1d/Input Channels:32/Output Channels:32/Kernel:3/Batch Size:2/Width:32/Optimized:0   |   12660235 ns   12639436 ns   |      55 |
| stateful_conv1d/Input Channels:32/Output Channels:32/Kernel:3/Batch Size:4/Width:64/Optimized:1    |   3265120 ns    3256765 ns     |   214 |
| stateful_conv1d/Input Channels:32/Output Channels:32/Kernel:3/Batch Size:4/Width:64/Optimized:0   |   33907082 ns   33729891 ns    |     21 |
| stateful_conv1d/Input Channels:32/Output Channels:32/Kernel:4/Batch Size:3/Width:128/Optimized:1   |   5152217 ns    5138281 ns    |    130 |
| stateful_conv1d/Input Channels:32/Output Channels:32/Kernel:4/Batch Size:3/Width:128/Optimized:0   |  69604856 ns   69372443 ns    |     10 |

kimishpatel commented:
I mean more side-to-side comparison would have been nice. So that you can just show what is the % improvement.

Commit:
Update on "Add benchmark for performance comparison"


Differential Revision: [D23197567](https://our.internmc.facebook.com/intern/diff/D23197567)

[ghstack-poisoned]

Commit:
Update on "Add benchmark for performance comparison"


Differential Revision: [D23197567](https://our.internmc.facebook.com/intern/diff/D23197567)

[ghstack-poisoned]

Commit:
Update on "Add benchmark for performance comparison"


Differential Revision: [D23197567](https://our.internmc.facebook.com/intern/diff/D23197567)

[ghstack-poisoned]

Commit:
Update on "Add benchmark for performance comparison"


Differential Revision: [D23197567](https://our.internmc.facebook.com/intern/diff/D23197567)

[ghstack-poisoned]
